SharedMemory/RpcBuffer.cs

```diff
@@ -890,130 +894,202 @@ bool WriteProtocolV1(MessageType msgType, ulong msgId, byte[] msg, ulong respons
 
             return true;
         }
-        
-        
-        
+
+        private Boolean m_ReadThreadIsReading = false;
+        private object m_ReadThreadIsReadingLock = new object();
+
+
 
         void ReadThreadV1()
         {
-            while(true && !ReadBuffer.ShuttingDown)
+            try
             {
-                if (Interlocked.Read(ref _disposed) == 1)
-                    return;
-
-                Statistics.StartWaitRead();
+                // Work with Local Variable to prefent NPE after dispose  
+                CircularBuffer l_TempReadBuffer = ReadBuffer;
 
-                ReadBuffer.Read((ptr) =>
+                while (true && !l_TempReadBuffer.ShuttingDown)
                 {
-                    int readLength = 0;
-                    var header = FastStructure<RpcProtocolHeaderV1>.PtrToStructure(ptr);
-                    ptr = ptr + protocolLength;
-                    readLength += protocolLength;
+                    if (Interlocked.Read(ref _disposed) == 1)
+                        return;
 
-                    RpcRequest request = null;
-                    if (header.MsgType == MessageType.RpcResponse || header.MsgType == MessageType.ErrorInRpc)
+                    // Check If Reading must be stopped
+                    if (_needDisposeManagedResources)
                     {
-                        if (!Requests.TryGetValue(header.ResponseId, out request))
-                        {
-                            // The response received does not have a  matching message that was sent
-                            Statistics.DiscardResponse(header.ResponseId);
-                            return protocolLength;
-                        }
+                        DisposeManagedResources();
+                        return;
                     }
-                    else
+
+                    // Set Marker for Reading in Progress
+                    lock (m_ReadThreadIsReadingLock)
```

Comment:
Anyway you can implement this without using locks? I'm concerned about how this may impact performance